src/System.Management.Automation/engine/parser/Compiler.cs

```diff
@@ -794,15 +797,49 @@ internal Expression ReduceAssignment(ISupportsAssignment left, TokenKind tokenKi
                 case TokenKind.MultiplyEquals: et = ExpressionType.Multiply; break;
                 case TokenKind.DivideEquals: et = ExpressionType.Divide; break;
                 case TokenKind.RemainderEquals: et = ExpressionType.Modulo; break;
+                case TokenKind.QuestionQuestionEquals when ExperimentalFeature.IsEnabled("PSNullCoalescingOperators"): et = ExpressionType.Coalesce; break;
             }
 
             var exprs = new List<Expression>();
             var temps = new List<ParameterExpression>();
             var getExpr = av.GetValue(this, exprs, temps);
-            exprs.Add(av.SetValue(this, DynamicExpression.Dynamic(PSBinaryOperationBinder.Get(et), typeof(object), getExpr, right)));
+
+            if(et == ExpressionType.Coalesce)
+            {
+                exprs.Add(av.SetValue(this, Coalesce(getExpr, right)));
+            }
+            else
+            {
+                exprs.Add(av.SetValue(this, DynamicExpression.Dynamic(PSBinaryOperationBinder.Get(et), typeof(object), getExpr, right)));
+            }
+
             return Expression.Block(temps, exprs);
         }
 
+        private Expression Coalesce(Expression left, Expression right)
+        {
+            Type leftType = left.Type;
+            Expression lhs = left.Cast(typeof(object));
+            Expression rhs = right.Cast(typeof(object));
+
+            if (leftType.IsValueType)
+            {
+                return lhs;
+            }
+            else if(leftType == typeof(DBNull) || leftType == typeof(NullString) || leftType == typeof(AutomationNull))
+            {
+                return rhs;
```

Comment:
Maybe we don't box the result in those two cases (`isValueType` and ` == DBNull || ...`)?

Reply:
Fixed